File: CSharp/Grouping/MinimumRatingGroupConfigurationSelector.cs
/*
 * Xceed DataGrid for WPF - SAMPLE CODE - Grouping Sample Application
 * Copyright (c) 2007-2024 Xceed Software Inc.
 * 
 * [MinimumRatingGroupConfigurationSelector.cs]
 *  
 * This class implements a GroupConfigurationSelector that
 * sets the ItemContainerStyle for the items whose Rating field is
 * at least MinimumRating when the DataGridControl is grouped by
 * Rating
 * 
 * This file is part of the Xceed DataGrid for WPF product. The use 
 * and distribution of this Sample Code is subject to the terms 
 * and conditions referring to "Sample Code" that are specified in 
 * the XCEED SOFTWARE LICENSE AGREEMENT accompanying this product.
 *
 */

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Windows;
using System.Windows.Data;
using System.Windows.Media;

namespace Xceed.Wpf.DataGrid.Samples.Grouping
{
  public class MinimumRatingGroupConfigurationSelector : GroupConfigurationSelector
  {

    #region PUBLIC CONSTRUCTORS

    public MinimumRatingGroupConfigurationSelector( int minimumRating )
    {
      this.MinimumRating = minimumRating;
    }

    #endregion

    #region PUBLIC PROPERTIES

    public int MinimumRating
    {
      get
      {
        return m_minItemCount;
      }
      set
      {
        if( value != m_minItemCount )
          m_minItemCount = value;
      }
    }

    #endregion

    #region PUBLIC METHODS

    public override GroupConfiguration SelectGroupConfiguration( int groupLevel,
                   System.Windows.Data.CollectionViewGroup collectionViewGroup,
                   System.ComponentModel.GroupDescription groupDescription )
    {
      GroupConfiguration groupConfiguration = null;

      if( collectionViewGroup == null )
        return base.SelectGroupConfiguration( groupLevel, collectionViewGroup, groupDescription );

      string fieldName = string.Empty;

      DataGridGroupDescription dataGridGroupDescription = groupDescription as DataGridGroupDescription;
      if( dataGridGroupDescription != null )
      {
        fieldName = dataGridGroupDescription.PropertyName;
      }
      else
      {
        PropertyGroupDescription propertyGroupDescription = groupDescription as PropertyGroupDescription;
        if( propertyGroupDescription != null )
        {
          fieldName = propertyGroupDescription.PropertyName;
        }
      }

      if( String.IsNullOrEmpty( fieldName ) == true )
        return base.SelectGroupConfiguration( groupLevel, collectionViewGroup, groupDescription );

      // We want this Selector to react only for "Rating" field
      if( fieldName.Equals( "Rating" ) == false )
        return base.SelectGroupConfiguration( groupLevel, collectionViewGroup, groupDescription );

      int rating = 0;

      if( Int32.TryParse( collectionViewGroup.Name.ToString(), out rating ) == true )
      {
        bool higherThan = ( rating >= this.MinimumRating );

        if( m_ratingHigherThanGroupConfiguration.ContainsKey( higherThan ) == true )
        {
          groupConfiguration = m_ratingHigherThanGroupConfiguration[ higherThan ];
        }
        else
        {
          Style style = new Style( typeof( Xceed.Wpf.DataGrid.DataRow ) );

          if( higherThan )
          {
            // the group's rating is high enough
            style.Setters.Add( new Setter( DataRow.BackgroundProperty, new SolidColorBrush( Color.FromArgb( 0x33, 0x00, 0x99, 0xcc ) ) ) );
          }
          else
          {
            // the group's rating is not high enough
            style.Setters.Add( new Setter( DataRow.BackgroundProperty, new SolidColorBrush( Color.FromArgb( 0x33, 0xff, 0x99, 0x00 ) ) ) );
          }

          // Create a new GroupConfiguration for this rating
          groupConfiguration = new GroupConfiguration();
          groupConfiguration.ItemContainerStyle = style;

          m_ratingHigherThanGroupConfiguration.Add( higherThan, groupConfiguration );
        }
      }
      else
      {
        Debug.WriteLine( "Error while converting rating to Int32" );
      }

      return groupConfiguration;
    }

    #endregion

    #region PRIVATE FIELDS

    private int m_minItemCount = 0;
    private Dictionary<bool, GroupConfiguration> m_ratingHigherThanGroupConfiguration = new Dictionary<bool, GroupConfiguration>();

    #endregion
  }
}

File: Visual Basic .NET/Grouping/MinimumRatingGroupConfigurationSelector.vb
'
' Xceed DataGrid for WPF - SAMPLE CODE - Grouping Sample Application
' Copyright (c) 2007-2024 Xceed Software Inc.
'
' [MinimumRatingGroupConfigurationSelector.vb]
'
' This class implements a GroupConfigurationSelector that
' sets the ItemContainerStyle for the items whose Rating field is
' at least MinimumRating when the DataGridControl is grouped by
' Rating
'
' This file is part of the Xceed DataGrid for WPF product. The use
' and distribution of this Sample Code is subject to the terms
' and conditions referring to "Sample Code" that are specified in
' the XCEED SOFTWARE LICENSE AGREEMENT accompanying this product.
'

Imports System
Imports System.Collections.Generic
Imports System.Text
Imports System.Windows.Media
Imports System.Windows
Imports System.Windows.Data
Imports System.Diagnostics
Imports System.ComponentModel

Namespace Xceed.Wpf.DataGrid.Samples.Grouping
  Public Class MinimumRatingGroupConfigurationSelector
    Inherits GroupConfigurationSelector

#Region "PUBLIC CONSTRUCTORS"

    Public Sub New(ByVal minimumRating As Integer)
      Me.MinimumRating = minimumRating
    End Sub

#End Region

#Region "PUBLIC PROPERTIES"

    Public Property MinimumRating() As Integer
      Get
        Return m_minItemCount
      End Get
      Set(ByVal value As Integer)
        If Value <> m_minItemCount Then
          m_minItemCount = Value
        End If
      End Set
    End Property

#End Region

#Region "PUBLIC METHODS"

    Public Overrides Function SelectGroupConfiguration(ByVal groupLevel As Integer, ByVal collectionViewGroup As System.Windows.Data.CollectionViewGroup, ByVal groupDescription As System.ComponentModel.GroupDescription) As GroupConfiguration
      Dim groupConfiguration As GroupConfiguration = Nothing

      If collectionViewGroup Is Nothing Then
        Return MyBase.SelectGroupConfiguration(groupLevel, collectionViewGroup, groupDescription)
      End If

      Dim fieldName As String = String.Empty

      Dim dataGridGroupDescription As DataGridGroupDescription = TryCast(groupDescription, DataGridGroupDescription)
      If Not dataGridGroupDescription Is Nothing Then
        fieldName = dataGridGroupDescription.PropertyName
      Else
        Dim propertyGroupDescription As PropertyGroupDescription = TryCast(groupDescription, PropertyGroupDescription)
        If Not propertyGroupDescription Is Nothing Then
          fieldName = propertyGroupDescription.PropertyName
        End If
      End If

      If String.IsNullOrEmpty(fieldName) = True Then
        Return MyBase.SelectGroupConfiguration(groupLevel, collectionViewGroup, groupDescription)
      End If

      ' We want this Selector to react only for "Rating" field
      If fieldName.Equals("Rating") = False Then
        Return MyBase.SelectGroupConfiguration(groupLevel, collectionViewGroup, groupDescription)
      End If

      Dim rating As Integer = 0

      If Int32.TryParse(collectionViewGroup.Name.ToString(), rating) = True Then
        Dim higherThan As Boolean = (rating >= Me.MinimumRating)

        If m_ratingHigherThanGroupConfiguration.ContainsKey(higherThan) = True Then
          groupConfiguration = m_ratingHigherThanGroupConfiguration(higherThan)
        Else
          Dim style As Style = New Style(GetType(Xceed.Wpf.DataGrid.DataRow))

          If higherThan Then
            ' the group's rating is high enough
            style.Setters.Add(New Setter(DataRow.BackgroundProperty, New SolidColorBrush(Color.FromArgb(&H33, &H00, &H99, &Hcc))))
          Else
            ' the group's rating is not high enough
			      style.Setters.Add(New Setter(DataRow.BackgroundProperty, New SolidColorBrush(Color.FromArgb(&H33, &Hff, &H99, &H00))))
          End If

          ' Create a new GroupConfiguration for this rating
          groupConfiguration = New GroupConfiguration()
          groupConfiguration.ItemContainerStyle = style

          m_ratingHigherThanGroupConfiguration.Add(higherThan, groupConfiguration)
        End If
      Else
        Debug.WriteLine("Error while converting rating to Int32")
      End If

      Return groupConfiguration
    End Function

#End Region

#Region "PRIVATE FIELDS"

    Private m_minItemCount As Integer = 0
    Private m_ratingHigherThanGroupConfiguration As Dictionary(Of Boolean, GroupConfiguration) = New Dictionary(Of Boolean, GroupConfiguration)()

#End Region
  End Class
End Namespace
